sa/sa.go

```diff
@@ -1748,6 +1749,48 @@ func (ssa *SQLStorageAuthority) RevokeCertificate(ctx context.Context, req *sapb
 	return &emptypb.Empty{}, nil
 }
 
+// UpdateRevokedCertificate stores new revocation information about an
+// already-revoked certificate. It will only store this information if the
+// cert is already revoked, if the new revocation reason is `KeyCompromise`,
+// and if the revokedDate is identical to the current revokedDate.
+func (ssa *SQLStorageAuthority) UpdateRevokedCertificate(ctx context.Context, req *sapb.RevokeCertificateRequest) (*emptypb.Empty, error) {
+	if req.Serial == "" || req.Date == 0 || req.Backdate == 0 || req.Response == nil {
+		return nil, errIncompleteRequest
+	}
+	if req.Reason != ocsp.KeyCompromise {
+		return nil, fmt.Errorf("cannot update revocation for any reason other than keyCompromise (1); got: %d", req.Reason)
+	}
+	thisUpdate := time.Unix(0, req.Date)
+	revokedDate := time.Unix(0, req.Backdate)
```

Comment:
This brings up an interesting problem. If the revocation date is updated in the database then it may be hard for external entities to see when the certificate was initially revoked for the first reason. This could cause compliance reports with claims that the certificate wasn't revoked within x hours of a request because they are going from the updated revocation date. This places the burden on the CA to dig up the previous OCSP response as proof.

Reply:
The idea here is that it _won't_ be updated: the SQL statement doesn't touch the revokedDate column, and the SQL statement will fail if the backdate/revokedDate contained in the request here is not equal to the value already in the database.

Reply:
Ah, yeah, I should have caught that. Thanks!